File: EDD/DLLPython/DoubleLinkedList.py
from NodoDoble import NodoDoble

class DoubleLinkedList:
    def __init__(self, value = None):
        if value == None:
            self.__size = 0
            self.__dll = None
        else:
            self.__dll = NodoDoble(value)
            self.__size = 1
    
    def getSize(self):
        return self.__size

    def isEmpty(self):
        if self.__dll == None:
            return True
        else:
            return False

    def append(self, value):
        if self.__dll == None:
            self.__dll = NodoDoble(value)
            self.__size += 1
        else:
            curr_node = self.__dll
            nuevo = NodoDoble(value)
            while(curr_node.getNext() != None):
                curr_node = curr_node.getNext()

            curr_node.setNext(nuevo)
            nuevo.setPrev(curr_node)
            self.__size += 1

    def findFromTail(self, value):
        curr_node = self.__dll
        posicion = 1
        #Llegar hasta el final de la lista
        while(curr_node.getNext() != None):
            curr_node = curr_node.getNext()
        
        #Ahora buscamos el valor
        while(curr_node != None):
            if(curr_node.getData() == value):
                return posicion
            curr_node = curr_node.getPrev()
            posicion += 1
        
        return None

    def findFromHead(self, value):
        curr_node = self.__dll
        posicion = 1
        while(curr_node != None):
            if(curr_node.getData() == value):
                return posicion
            
            curr_node = curr_node.getNext()
            posicion += 1
        
        return None

    def removeFromTail(self, value):
        curr_node = self.__dll
        #Llegar hasta el final de la lista
        while(curr_node.getNext() != None):
            curr_node = curr_node.getNext()
        
        #Ahora buscamos el valor
        while(curr_node != None):
            if(curr_node.getData() == value):
                
                if(curr_node.getNext() == None):
                    aux = curr_node.getPrev()
                    aux.setNext(None)
                    curr_node.setPrev(None)
                    self.__size -= 1
                elif(curr_node.getPrev() == None):
                    aux = curr_node.getNext()
                    aux.setPrev(None)
                    curr_node.setNext(None)
                    self.__dll = aux
                    self.__size -= 1
                else:
                    aux = curr_node.getPrev()
                    auxDos = curr_node.getNext()
                    
                    curr_node.setNext(None)
                    curr_node.setPrev(None)
                    
                    aux.setNext(auxDos)
                    auxDos.setPrev(aux)
                    self.__size -= 1
            
            curr_node = curr_node.getPrev()

    def removeFromHead(self, value):
        curr_node = self.__dll
        while(curr_node != None):
            if(curr_node.getData() == value):
                
                if(curr_node.getNext() == None):
                    aux = curr_node.getPrev()
                    aux.setNext(None)
                    curr_node.setPrev(None)
                    self.__size -= 1
                elif(curr_node.getPrev() == None):
                    aux = curr_node.getNext()
                    aux.setPrev(None)
                    curr_node.setNext(None)
                    self.__dll = aux
                    self.__size -= 1
                else:
                    aux = curr_node.getPrev()
                    auxDos = curr_node.getNext()
                    
                    curr_node.setNext(None)
                    curr_node.setPrev(None)
                    
                    aux.setNext(auxDos)
                    auxDos.setPrev(aux)
                    self.__size -= 1
                
            curr_node = curr_node.getPrev()

    def transversal(self):
        curr_node = self.__dll;
        while(curr_node != None):
            print(f"[{curr_node.getData()}]", end="")
            curr_node = curr_node.getNext()
        
        print("")

    def reverseTransversal(self):
        curr_node = self.__dll
        #Llegar hasta el final de la lista
        while(curr_node.getNext() != None):
            curr_node = curr_node.getNext()
        
        #Ya en el ultimo elemento hacemos el recorrido al reves
        while(curr_node != None):
            print(f"[{curr_node.getData()}]", end="")
            curr_node = curr_node.getPrev()

        print()

File: EDD/DLL/src/main/java/ico/fes/unam/edd1358/DoubleLinkedList.java
package ico.fes.unam.edd1358;

public class DoubleLinkedList {
    
    private class NodoDoble<T> {
        private T data;
        private NodoDoble next = null;
        private NodoDoble prev = null;

        public NodoDoble() {
        }

        public NodoDoble(T data) {
            this.data = data;
        }

        public NodoDoble(T data, NodoDoble siguiente, NodoDoble anterior){
            this.data = data;
            this.next = siguiente;
            this.prev = anterior;
        }

        public T getData() {
            return data;
        }

        public void setData(T data) {
            this.data = data;
        }

        public NodoDoble getNext() {
            return next;
        }

        public void setNext(NodoDoble next) {
            this.next = next;
        }

        public NodoDoble getPrev() {
            return prev;
        }

        public void setPrev(NodoDoble prev) {
            this.prev = prev;
        }

    }
    
    private NodoDoble head;
    private NodoDoble tail;
    private int size;
    
    public DoubleLinkedList(){
        this.size = 0;
    }
    
    public DoubleLinkedList(Object value){
        this.head = new NodoDoble(value);
        this.tail = this.head;
        this.size = 1;
    }
    
    public int getSize(){
        return this.size;
    }
    
    public boolean isEmpty(){
        return this.head == null;
    }
    
    public void append(Object value){
        if(this.isEmpty()){
            this.head = new NodoDoble(value);
            this.tail = this.head;
        } else {
            this.tail.setNext(new NodoDoble(value, null, this.tail));
            this.tail = this.tail.getNext();
        }
        this.size ++;
    }
    
    public NodoDoble findFromTail(Object value){
        NodoDoble currNode = this.tail;
        //Buscamos el valor
        while(currNode != null){
            if(currNode.getData().toString().equals(value.toString())){
                return currNode;
            }
            currNode = currNode.getPrev();
        }
        return null;
    }
    
    public NodoDoble findFromHead(Object value){
        NodoDoble currNode = this.head;
        while(currNode != null){
            if(currNode.getData().toString().equals(value.toString())){
                return currNode;
            }
            currNode = currNode.getNext();
        }
        return null;
    }
    
    public void removeFromTail(Object value){
        NodoDoble currNode = this.tail;
        //Buscamos el valor
        while(currNode != null){
            if(currNode.getData().toString().equals(value.toString())){
                
                if(currNode.getNext() == null){
                    NodoDoble aux = currNode.getPrev();
                    aux.setNext(null);
                    currNode.setPrev(null);
                    System.out.println("Eliminado correctamente.");
                    this.size --;
                } else if(currNode.getPrev() == null){
                    NodoDoble aux = currNode.getNext();
                    aux.setPrev(null);
                    currNode.setNext(null);
                    this.head = aux;
                    System.out.println("Eliminado correctamente.");
                    this.size --;
                } else {
                    NodoDoble aux = currNode.getPrev();
                    NodoDoble auxDos = currNode.getNext();
                    
                    currNode.setNext(null);
                    currNode.setPrev(null);
                    
                    aux.setNext(auxDos);
                    auxDos.setPrev(aux);
                    System.out.println("Eliminado correctamente.");
                    this.size --;
                }
            }
            currNode = currNode.getPrev();
        }
        this.updateTail();
    }
    
    public void removeFromHead(Object value){
        NodoDoble currNode = this.head;
        while(currNode != null){
            if(currNode.getData().toString().equals(value.toString())){
                
                if(currNode.getNext() == null){
                    NodoDoble aux = currNode.getPrev();
                    aux.setNext(null);
                    currNode.setPrev(null);
                    System.out.println("Eliminado correctamente.");
                    this.size --;
                } else if(currNode.getPrev() == null){
                    NodoDoble aux = currNode.getNext();
                    aux.setPrev(null);
                    currNode.setNext(null);
                    this.head = aux;
                    System.out.println("Eliminado correctamente.");
                    this.size --;
                } else {
                    NodoDoble aux = currNode.getPrev();
                    NodoDoble auxDos = currNode.getNext();
                    
                    currNode.setNext(null);
                    currNode.setPrev(null);
                    
                    aux.setNext(auxDos);
                    auxDos.setPrev(aux);
                    
                    System.out.println("Eliminado correctamente.");
                    this.size --;
                }
            }
            currNode = currNode.getPrev();
        }
        this.updateTail();
    }
    
    public void transversal(){
        NodoDoble currNode = this.head;
        while(currNode != null){
            System.out.print("[\n" + currNode.getData().toString() + "\n], ");
            currNode = currNode.getNext();
        }
        System.out.println("");
    }
    
    public void reverseTransversal(){
        NodoDoble currNode = this.tail;
        //Ya en el ultimo elemento hacemos el recorrido al reves
        while(currNode != null){
            System.out.print("[\n" + currNode.getData().toString() + "\n], ");
            currNode = currNode.getPrev();
        }
        System.out.println("");
    }
    
    private void updateTail(){
        NodoDoble currNode = this.head;
        while(currNode.getNext() != null){
            currNode = currNode.getNext();
        }
        this.tail = currNode;
    }
    
}
